kafka/src/main/java/nl/nn/adapterframework/extensions/kafka/KafkaListener.java

```diff
@@ -0,0 +1,142 @@
+/*
+   Copyright 2023 WeAreFrank!
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.adapterframework.extensions.kafka;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nonnull;
+
+import lombok.AccessLevel;
+
+import nl.nn.adapterframework.stream.MessageContext;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+import lombok.Setter;
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.core.IPullingListener;
+import nl.nn.adapterframework.core.ListenerException;
+import nl.nn.adapterframework.core.PipeLineResult;
+import nl.nn.adapterframework.core.PipeLineSession;
+import nl.nn.adapterframework.receivers.RawMessageWrapper;
+import nl.nn.adapterframework.stream.Message;
+
+public class KafkaListener extends KafkaFacade implements IPullingListener<ConsumerRecord<String, byte[]>> {
+	//setter is for testing purposes only.
+	private @Setter(AccessLevel.PACKAGE) Consumer<String, byte[]> consumer;
+	private final List<ConsumerRecord<String, byte[]>> waiting = new ArrayList<>();
```

Comment:
Why do you make this a `List<>` instead of an `Iterator<>` which you can directly obtain from the `ConsumerRecords<>` instance which you get from the `poll()` method?

Then you do not have to copy, do not have to remove first element from list all the time, and can just return `iterator.next()` for as long as `iterator.hasNext()`.